cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/classloader/SparkClassRewriter.java

```diff
@@ -663,6 +672,95 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
+   */
+  private byte[] rewriteTempFileNameForCheckpoint(InputStream byteCodeStream) throws IOException {
+    String checkpointTimeFieldName =
+      "org$apache$spark$streaming$CheckpointWriter$CheckpointWriteHandler$$checkpointTime";
+    Type timeType = Type.getType(org.apache.spark.streaming.Time.class);
+    Type hadoopPathType = Type.getType(org.apache.hadoop.fs.Path.class);
+    Type stringType = Type.getType(String.class);
+    Type stringBuilderType = Type.getType(StringBuilder.class);
+
+    ClassReader cr = new ClassReader(byteCodeStream);
+    ClassWriter cw = new ClassWriter(0);
+
+    cr.accept(new ClassVisitor(Opcodes.ASM5, cw) {
+      @Override
+      public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
+
+        MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+
+        return new GeneratorAdapter(Opcodes.ASM5, mv, access, name, desc) {
+
+          boolean tempStringAddedToStack = false;
+          final Method hadoopPathConstructorMethod =
+            new Method("<init>", Type.VOID_TYPE, new Type[]{ stringType, stringType });
+
+          @Override
+          public void visitLdcInsn(Object value) {
+            // If we find this instruction, we update our atomic boolean to True:
+            // LDC "temp"
+            // If this value is true, and the next instruction is the Path constructor, we need to
+            // concat the checkpoont time to this string
+            tempStringAddedToStack = "temp".equals(value);
+
+            super.visitLdcInsn(value);
+          }
+
+          @Override
+          public void visitMethodInsn(int opcode, String owner, String name, String desc, boolean itf) {
+
+            // If we see a call to the Path constructor with 2 String parameters, rewrite the second argument
+            // to add a timestamp.
+            if (opcode == Opcodes.INVOKESPECIAL
+              && Type.getObjectType(owner).equals(hadoopPathType)
+              && hadoopPathConstructorMethod.equals(new Method(name, desc))
+              && tempStringAddedToStack) {
+
+              //Pop the LOC "temp" argument from the stack.
+              //super.visitInsn(Opcodes.POP);
+              pop();
+
+              //Build the string that replaces "temp" as "temp-" + System.currentTimestampMillis.
+              //This function generates the following code:
+              //  new StringBuilder().append("temp-").append(checkpointTime.milliseconds()).toString();
+              newInstance(stringBuilderType);
+              dup();
+              invokeConstructor(stringBuilderType, Methods.getMethod(void.class, "<init>"));
+              push("temp-");
+              invokeVirtual(stringBuilderType,
+                            new Method("append", stringBuilderType, new Type[]{stringType}));
+              loadThis();
+              getField(CHECKPOINT_WRITE_HANDLER_TYPE, checkpointTimeFieldName, timeType);
```

Comment:
You need to override the visitField  of the ClassWriter to make sure this class actually has this field. Otherwise you can result in cryptic classloading failure

Reply:
Done!